### PR TITLE
Backend adds the ability to support Redis clusters

### LINKED ISSUE #1555: Support the Redis cluster on the backend
### What would you like to be added?

The backend supports the Redis cluster capability.

### Why is this needed?

Currently, the backend supports only the Redis single-node system but not the Redis cluster. To facilitate user operations, the Redis cluster capability is supported.

### PR DESCRIPTION
**Pull request by hanbingleixue**
**What type of PR is this?**

Feature.

**What this PR does / why we need it?**

Describe it.

1、Backend adds the ability to support Redis clusters

**Which issue(s) this PR fixes？**

Fixes #1555 

**Does this PR introduce a user-facing change?**

No

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/sermant-io/Sermant/issues) related with this PR before you start working on it.
- [x] Make sure you have squashed your change to one single commit.
- [x] GitHub Actions works fine in this PR.

### CODE REVIEW

File: sermant-backend/src/main/java/io/sermant/backend/dao/redis/EventDaoForRedis.java [hunk 1]
```diff
@@ -49,6 +51,11 @@ public class EventDaoForRedis implements EventDao {
      */
     public EventDaoForRedis(BackendConfig backendConfig) {
         try {
+            String[] addressList = backendConfig.getUrl().split(CommonConst.REDIS_CLUSTER_SPLIT);
+            if (addressList.length > 1) {
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/dao/redis/EventDaoForRedis.java]: trim before split

File: sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (C) 2023-2023 Huawei Technologies Co., Ltd. All rights reserved.
+ *
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java]: update copyright

File: sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+import redis.clients.jedis.JedisClientConfig;
+import redis.clients.jedis.JedisCluster;
+import redis.clients.jedis.resps.Tuple;
+
+/**
+ * Redis stand-alone client
+ *
+ * @author xuezechao
+ * @since 2023-03-02
+ */
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java]: comment not correct

File: sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    private JedisCluster jedisCluster;
+
+    /**
+     * Construct the Jedis connection pool
+     *
+     * @param backendConfig configuration
+     */
+    public RedisClusterClientImpl(BackendConfig backendConfig) {
+        this.backendConfig = backendConfig;
+        String[] addressList = backendConfig.getUrl().split(CommonConst.REDIS_CLUSTER_SPLIT);
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java]: no need to process addressList again, put it into RedisClusterClientImpl constructor

File: sermant-backend/src/main/java/io/sermant/backend/server/EventServer.java [hunk 5]
```diff
@@ -64,10 +66,15 @@ private EventServer() {
...
-            eventServer.daoService = new EventDaoForRedis(backendConfig);
-        } else {
-            eventServer.daoService = new MemoryClientImpl(backendConfig);
+            String[] addressList = backendConfig.getUrl().split(CommonConst.REDIS_CLUSTER_SPLIT);
+            if (addressList.length > 1) {
+                eventServer.daoService = new RedisClusterClientImpl(backendConfig);
+                return;
+            }
+            eventServer.daoService = new RedisClientImpl(backendConfig);
+            return;
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/server/EventServer.java]: RedisClusterClientImpl already support stand-alone mode

File: sermant-backend/src/main/java/io/sermant/backend/dao/redis/EventDaoForRedis.java (hunk 1 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/dao/redis/EventDaoForRedis.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java (hunk 2 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java (hunk 3 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java (hunk 4 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisClusterClientImpl.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/server/EventServer.java (hunk 5 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/server/EventServer.java]: RedisClusterClientImpl can be connected only after the Redis cluster configuration is enabled.

File: sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisOperation.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+ * @since 2024-06-25
+ */
+public interface RedisOperation {
+    /**
+     * Get the value of the specified key
+     *
+     * @param pattern key
+     * @return the value stored at key
+     */
+    Set<String> keys(String pattern);
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisOperation.java:37]: getValue？

Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/dao/redis/RedisOperation.java:37]: done

**Review by lilai23 — APPROVED** [review]

**Review by daizhenyu — APPROVED** [review]

**Pull request merged**